App_Use_Cases.tex

```diff
@@ -502,7 +508,85 @@ \subsection{Use Case Details}
+
+Cross-version compatibility responsibilities are not restricted to \ac{PMIx}, but a general issue for any library that coordinates across multiple processes. This includes, but not limited to, client/server libraries, and libraries with a user-space and kernel-space component (e.g., high-performance interconnect libraries).
+
+\subsection{Use Case Details}
+
+There are three scenarios that a \ac{PMIx} implementation and a \ac{PMIx} consumer must consider.
+These scenarios use a PMIx Server and a PMIx Client for clarity, though the scenarios also apply to PMIx Tools.
+\begin{enumerate}
+\item \textbf{PMIx Server version matches PMIx Client version}:
+No cross-version considerations are necessary since they are running the same version.
+\item \textbf{PMIx Server version is older than PMIx Client version}:
+The implementation must negotiate capabilities during the initial handshake.\\
+This scenario is common if the (possibly containerized) PMIx client application is being run on an established system that does not update as frequently as the application requires.
+Thus the PMIx Server in the \ac{RM} is locked to an older version of the \ac{PMIx} implementation.
+\item \textbf{PMIx Server version is newer than PMIx Client version}:
+The implementation must negotiate capabilities during the initial handshake.\\
+This scenario is common if the (possibly containerized) PMIx client application is being run after a system software upgrade on the system.
+Thus the PMIx Server in the \ac{RM} has been upgraded to a newer version of the \ac{PMIx} implementation and the client is still linked against the older version.
+\end{enumerate}
+
+When the two \ac{PMIx}-enabled processes first connect to each other they need to first check the version of the library that they are each running.
+This handshake often occurs during initialization (though it could occur on a per-operation basis depending on the \ac{PMIx} implementation), for example during the following operations:
+\begin{itemize}
+\item PMIx Clients: \refapi{PMIx_Init}
+\item PMIx Tools: \refapi{PMIx_tool_init}, \refapi{PMIx_tool_attach_to_server}
+\item PMIx Servers: \refapi{PMIx_server_init}, \refapi{pmix_server_client_connected2_fn_t}, \refapi{pmix_server_tool_connection_fn_t}
+\end{itemize}
+
+Commonly this cross-version handshake occurs completely transparently to the consumers of the \ac{PMIx} interface since it happens inside the \ac{PMIx} implementation of these interfaces.
+However, during the negotiation, some features available in one version might not be available in the other.
+The consumer of the \ac{PMIx} interface should always be prepared to receive the \refconst{PMIX_ERR_NOT_SUPPORTED} error code from a \ac{PMIx} interface call that the other side either does not support or is not available in the version of the library with which they are linked.
```

Comment:
Presumably forward-compatibility here is an issue when a newer client tries to call a function/interface that doesn't exist in an older server (since the linker will fail)?  But a newer client passing in an attribute that doesn't exist in the older server would probably work and the client would (probably?) get a `PMIX_ERR_NOT_SUPPORTED`?  If so, might be good to call out that difference in the interface vs attribute forwards compatibility.

Reply:
It doesn't quite work that way. The notion of cross-compatibility implies that the client is built and linked against one version of the library, and the server is built and linked against a different one. In your example, the newer client is able to build and link against the newer version of the library because the new API is defined there. When they call the new API, the library will send a request up to the server. The older version of the server library is required to return a `PMIX_ERR_NOT_SUPPORTED` value since it won't recognize the request - it specifically is _not_ allowed to segfault or silently ignore the request.

Attributes work the same way. The newer client likely won't have an issue with an older attribute, but the older server must cleanly handle any newer attribute - ignoring it if it isn't required, returning "not supported" if the attribute is marked as required.